microprofile/WFLY-11529_subsystem_metrics.asciidoc

```diff
@@ -0,0 +1,145 @@
+
+[NOTE]
+====
+* the `exposed-subsystems` was already present in the subsystem but it had no impact since the WildFly metrics from subsystems were disabled in https://issues.jboss.org/browse/WFLY-11399[WFLY-11399] after it was found out that the first implementation would not allow aggregation of these metrics.
+With this feature, the `exposed-subsystems` will be relevant again.
+====
+
+==== Metric name and labels
+
+The name of a WildFly metric is based on the subsystem that provides it as well as the name of the attribute from the management model.
+The name can also be prepended with a `prefix` (specified on the `/subsystem=microprofile-metrics` resource).
+The name can also be appended with the `_total` suffix if it corresponds to a Counter metrics (as explained in https://issues.jboss.org/browse/WFCORE-4238[WFCORE-4238])
+Other information are stored using labels.
+
+For example Undertow exposes a metric attribute `request-count` (which is a counter) for every Servlet in an application deployment.
+This attribute will be exposed to Prometheus with the name `wildfly_undertow_request_count_total`.
+Other information such as the name of the Servlet are added to the labels of the metric.
+
+The https://github.com/wildfly/quickstart/tree/master/helloworld-rs[helloworld-rs quickstart] creates a JAX-RS application that can be deployed in WildFly.
+ A corresponding metric will be exposed for it with the name and labels:
+
+----
+wildfly_undertow_request_count_total{deployment="helloworld-rs.war",servlet="org.jboss.as.quickstarts.rshelloworld.JAXActivator",subdeployment="helloworld-rs.war"}
+----
+
+For deployment metrics (as above), if the metric belongs to a `deployment`, the `subdeployment` label is also added (with the `deployment` label value) in
+order to facilitate aggregation of data by Prometheus.
+
+The example below returns 3 counters for Undertow servlet request count:
+* for the HTTP and HTTPS listeners in the `undertow` subsystem
+* for one servlet from the https://github.com/wildfly/quickstart/tree/master/helloworld-rs[helloworld-rs quickstart] deployment:
+----
+$ curl -v http://127.0.0.1:9990/metrics/ | grep wildfly_undertow_request_count_total
+...
+# HELP wildfly_undertow_request_count_total The number of requests this listener has served
+# TYPE wildfly_undertow_request_count counter
+wildfly_undertow_request_count_total{server="default-server",http_listener="default",} 8.0
+wildfly_undertow_request_count_total{server="default-server",https_listener="https",} 0.0
+wildfly_undertow_request_count_total{deployment="helloworld-rs.war",servlet="org.jboss.as.quickstarts.rshelloworld.JAXActivator",subdeployment="helloworld-rs.war",} 6.0
+...
+----
+
+===== Implementation Issues
+
+====== Complex WildFly Metrics Are Not Supported
+
+Resources may return arbitrarily complex return types for metrics.
+The registration code in `microprofile-smallrye-metrics` will only register WildFly metrics for simple numerical `ModelType`.
```

Comment:
Is there a warning when such complex type is encountered ?

Reply:
no, there is no warning. We do not need to warn the user as there is nothing he could do about it.
We have some metrics that returns STRING (e.g. enum of server state) which is correct from WildFly point of view.
However Prometheus only consumes numerical metric.

If/When we decide to support more complex metrics (eg histogram), we would have to create corresponding RFEs for it.